### PR TITLE
Implement WaNet Backdoor Transform

### PR DESCRIPTION
**Pull request by VRehnberg**
This implements a `Transform` for the WaNet backdoor described in
> Nguyen, T. A., & Tran, A. T. (2021). WaNet - Imperceptible Warping-based Backdoor Attack. International Conference on Learning Representations. Retrieved from https://openreview.net/forum?id=eEn8KTtJOx

### CODE REVIEW

**Comment by ejnnr** [comment]
Thanks, this looks really good at a first glance! (Haven't carefully read over the implementation yet, but the high-level structure is great)

I don't know whether you've decided you want to spend more time on this, no obligations obviously. If you do, two small general things:
1. I'm using some (not yet documented) precommit hooks for formatting, linting, etc. You can set it up by `pip install pre-commit`, `pre-commit install` (to set up git hooks) and `pre-commit run --all-files` to make sure it runs against changes from your past commits here. Usually any linting issues should be fixed automatically, but some (like line length in strings) need to be fixed manually.
2. Would be nice to have this in `tests/test_pipeline.py` (which basically just tests that the scripts run without errors). Easiest option would be to just use the new WaNet backdoor instead of the pixel corner one (which is simpler and I'm less likely to break it accidentally with some future change). Ideally we'd of course have a more comprehensive test suite at some point but that doesn't need to happen in this PR.

Also, did you have any way to at least informally check the implementation does what you'd expect/matches the paper? I guess the most important thing is just that networks can learn the backdoor, and that it's imperceptible to humans, though ideally we'd have some way of getting even more certainty.

Btw, for some reason I didn't get an email notification for this PR, so if I'm slow to respond to new PRs or other stuff on Github in the future, please feel free to ping me by email!

**Comment by VRehnberg** [comment]
> I don't know whether you've decided you want to spend more time on this

I've decided to work on this for now. We'll have to talk some more about goals at some point but this issue is probably not the right place.

1. A contributing instructions would probably be a good idea in the future. I'll collect what I learn and make a PR when I've got something more substantial.
2. I did this locally before in all places before I pushed, I'll add one of the tests upstream as well.

> Also, did you have any way to at least informally check the implementation does what you'd expect/matches the paper? I guess the most important thing is just that networks can learn the backdoor, and that it's imperceptible to humans, though ideally we'd have some way of getting even more certainty.

Nothing that fancy. I looked a lot at images and how they where affected by the transform and diffs between before and after transformation. All I can tell is that the transform is mostly imperceptible, you can tell that straight lines are sometimes curved slightly if look closely and that these changes match the warping field.

I'll compare it side by side with the original paper's implementation in the future.

> Btw, for some reason I didn't get an email notification for this PR, so if I'm slow to respond to new PRs or other stuff on Github in the future, please feel free to ping me by email!

Absolutely, will do so if I need a faster response.

**Comment by ejnnr** [comment]
> I've decided to work on this for now. We'll have to talk some more about goals at some point but this issue is probably not the right place.

Happy to hear that of course! Feel free to email me to set up more meetings to talk about next steps.

> A contributing instructions would probably be a good idea in the future. I'll collect what I learn and make a PR when I've got something more substantial.

I've added a few more instructions in the README and the [docs](https://github.com/ejnnr/cupbearer/blob/main/docs/getting_started.md) now but it's probably still incomplete, so additions would be great!

> Nothing that fancy. I looked a lot at images and how they where affected by the transform and diffs between before and after transformation. All I can tell is that the transform is mostly imperceptible, you can tell that straight lines are sometimes curved slightly if look closely and that these changes match the warping field.

Ok, that sounds good to me, don't think it really makes sense to do anything more complicated.

Looks to me like this PR is (basically?) finished. Let me know whenever you're done and I'll take a final more detailed look.

**Comment by VRehnberg** [comment]
So I've compared it more in detail with the [original implementation](https://github.com/VinAIResearch/Warping-based_Backdoor_Attack-release/blob/main/train.py) and found a few differences:
 - [x] With a given warping tensor x, y axes are flipped. But this doesn't matter as they are both sampled identically distributed.
 - [ ] I haven't implemented the grid_rescale hyperparameter other than implicitly using the default 1.0.
   - It should be easily done so might just as well do it.
 - [ ] I do not denormalize the input image. I don't see how this could impact the transform and it doesn't seem to be mentioned in the paper. On the other hand, I don't think they would have gone through the effort of doing that unless it mattered.
 - [ ] The warping strength differs with a factor of two. `grid_sample` which is used in the original implementation and BackdoorBench uses normalized coordinates in [`-1`, `1`] instead of [`0`, `img.shape[dim]`]. However, when going from relative coordinates they divide by `img.shape[dim]` instead of fully changing units. Possible solutions:
     - Simply do what the original implementation does to be consistent. Easily done, but strange unit.
     - Change the default value by a factor of two and simply note how the units differ from the original implementation.
 - [ ] `map_coordinates(order=3)` and torch's `interpolate`/`upsample("bicubic")` are slightly different. RMSE is a few percent of the mean pixel value in the upsampling step for `np.rand.random((4,4))` to size (28, 28).

**Comment by ejnnr** [comment]
Thanks for this list!

> I do not denormalize the input image. I don't see how this could impact the transform and it doesn't seem to be mentioned in the paper. On the other hand, I don't think they would have gone through the effort of doing that unless it mattered.

Are you referring to this line in the original? https://github.com/VinAIResearch/Warping-based_Backdoor_Attack-release/blob/94453080f241053ac7c8cc4717da20806ee17e5c/train.py#L149
That looks like it's only for visualizing the images in tensorboard. I currently don't normalize images anyway (should clearly add that at some point). But even if we have normalization, denormalizing for visualization should be done elsewhere.

> The warping strength differs with a factor of two. grid_sample which is used in the original implementation and BackdoorBench uses normalized coordinates in [-1, 1] instead of [0, img.shape[dim]]. However, when going from relative coordinates they divide by img.shape[dim] instead of fully changing units. Possible solutions:

Could we just add something like `warping_strength *= 2` (or `/= 2`) somewhere and not mess with the indexing? As far as I can tell, the warping strength is on a somewhat arbitrary scale anyway (i.e. warping strength = 1 doesn't have an especially important meaning). But correct me if I'm wrong about that! Assuming that's the case, I think it makes sense to have our warping strength mean the same thing as theirs, but if possible stay with the more common [0, n] indexing.

> map_coordinates(order=3) and torch's interpolate/upsample("bicubic") are slightly different. RMSE is a few percent of the mean pixel value in the upsampling step for np.random.rand((4,4)) to size (28, 28).

I would hope this kind of things doesn't matter for the method, so I think it's fine to ignore for now. Maybe just add a comment saying this is slightly different, so we know it's a thing to look into if we run into problems in the future. (Alternatively, we could just use `torch`'s `upsample` here since we're relying on pytorch already---conversion to and from numpy arrays should be free IIRC. I don't have a strong take either way).

**Comment by VRehnberg** [comment]
> Are you referring to this line in the original? https://github.com/VinAIResearch/Warping-based_Backdoor_Attack-release/blob/94453080f241053ac7c8cc4717da20806ee17e5c/train.py#L149
> That looks to me like it's only for visualizing the images in tensorboard. I currently don't normalize images anyway (should clearly add that at some point). But even if we have normalization, denormalizing for visualization should be done elsewhere.

For this comment I was probably mostly looking at [this line](https://github.com/SCLBD/BackdoorBench/blob/main/attack/wanet.py#L294) in BackdoorBench, but noting that they did something with denormalize in the original as well. But, as long as it doesn't matter in the original it should be fine.

**Comment by VRehnberg** [comment]
> Could we just add something like `warping_strength *= 2` (or `/= 2`) somewhere

Yes. I'll do that.

> As far as I can tell, the warping strength is on a somewhat arbitrary scale anyway (i.e. warping strength = 1 doesn't have an especially important meaning).

That's about right. The relevant steps of the transform are
1. Sample a small grid from a Uniform distribution
2. Normalize this so that the mean absolute value is 1*warping strength
3. Smooth this out into the same shape as the image
4. Sample each pixel by the offset given from the smoothed out grid

So I believe the two different meanings of warping strength 1 would be:
 - absolute pixel offset is on average 1 in vertical & horizontal direction
 - the expected maximum absolute pixel offset is 1 in vertical & horizontal direction

Both of these are should be valid units.

**Comment by VRehnberg** [comment]
> I would hope this kind of things doesn't matter for the method, so I think it's fine to ignore for now. Maybe just add a comment saying this is slightly different, so we know it's a thing to look into if we run into problems in the future. (Alternatively, we could just use `torch`'s `upsample` here since we're relying on pytorch already---conversion to and from numpy arrays should be free IIRC. I don't have a strong take either way).

I'll do the comment for now. I prefer not to mix frameworks in the middle of computations.

**Comment by VRehnberg** [comment]
I think this should be ready to be merged now.
